fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,140 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    const myRequest = new Request('/api/todos', {
+      method: 'GET'
+    });
+
+    fetch(myRequest)
+      .then(res => {
+        return res.json()
+      })
+      .then(items => {
+        console.log(items);
+        this.setState({
+          todos: items
+        })
+      })
+      .catch(console.error);
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    const myRequest = new Request('/api/todos', {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newTodo)
+    });
+
+    fetch(myRequest)
+      .then(res => {
+        return res.json()
+      })
+      .then(items => {
+        let newArray = this.state.todos;
```

Comment:
This wont make a copy of the array, it just creates a reference to the array, so on the next line with `push` you will mutate the state. To make a shallow copy of the array you can do it 1 of 2 ways:

ES6 Spread operator:
```JAVASCRIPT
const newArray = [ ...this.state.todos ];
```

OR

Using the `slice` method
```JAVASCRIPT
const newArray = this.state.todos.slice();
```

Reply:
Thanks and updated!